File: sdk/go/validate/validate.go
package validate

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
	"sync"

	_ "embed"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

//go:embed schemas/openrtb.schema.json
var schemaOpenRTB []byte

//go:embed schemas/bid-request.schema.json
var schemaBidRequest []byte

//go:embed schemas/bid-response.schema.json
var schemaBidResponse []byte

//go:embed schemas/native.schema.json
var schemaNative []byte

var (
	compilerOnce sync.Once
	compilerErr  error
	requestSch   *jsonschema.Schema
	responseSch  *jsonschema.Schema
	nativeSch    *jsonschema.Schema
)

func initSchemas() {
	compilerOnce.Do(func() {
		c := jsonschema.NewCompiler()
		resources := map[string][]byte{
			"openrtb.schema.json":      schemaOpenRTB,
			"bid-request.schema.json":  schemaBidRequest,
			"bid-response.schema.json": schemaBidResponse,
			"native.schema.json":       schemaNative,
		}
		for name, raw := range resources {
			doc, err := jsonschema.UnmarshalJSON(bytes.NewReader(raw))
			if err != nil {
				compilerErr = fmt.Errorf("%s: %w", name, err)
				return
			}
			if err := c.AddResource(name, doc); err != nil {
				compilerErr = fmt.Errorf("%s: %w", name, err)
				return
			}
			if m, ok := doc.(map[string]any); ok {
				if id, ok := m["$id"].(string); ok && id != "" {
					if err := c.AddResource(id, doc); err != nil {
						compilerErr = fmt.Errorf("%s ($id): %w", name, err)
						return
					}
				}
			}
		}
		var err error
		requestSch, err = c.Compile("bid-request.schema.json")
		if err != nil {
			compilerErr = err
			return
		}
		responseSch, err = c.Compile("bid-response.schema.json")
		if err != nil {
			compilerErr = err
			return
		}
		nativeSch, err = c.Compile("native.schema.json")
		if err != nil {
			compilerErr = err
			return
		}
	})
}

// ValidateBidRequest 校验 BidRequest JSON 字节。
func ValidateBidRequest(data []byte) ValidationResult {
	initSchemas()
	return validateKind(data, requestSch, true)
}

// ValidateBidResponse 校验 BidResponse JSON 字节。
func ValidateBidResponse(data []byte) ValidationResult {
	initSchemas()
	return validateKind(data, responseSch, false)
}

func validateKind(data []byte, schema *jsonschema.Schema, checkNative bool) ValidationResult {
	if compilerErr != nil {
		return Fail(ValidationError{Code: "constraint", Path: "", Message: compilerErr.Error()})
	}
	if schema == nil {
		return Fail(ValidationError{Code: "constraint", Path: "", Message: "schema not loaded"})
	}
	var doc any
	if err := json.Unmarshal(data, &doc); err != nil {
		return Fail(ValidationError{Code: "parse", Path: "", Message: err.Error()})
	}
	if err := schema.Validate(doc); err != nil {
		return Fail(mapSchemaError(err)...)
	}
	if checkNative {
		if errs := validateNativeEmbedded(doc); len(errs) > 0 {
			return Fail(errs...)
		}
	}
	return OK()
}

func mapSchemaError(err error) []ValidationError {
	var out []ValidationError
	switch e := err.(type) {
	case *jsonschema.ValidationError:
		collectVE(e, &out)
	default:
		out = append(out, ValidationError{Code: "constraint", Path: "", Message: err.Error()})
	}
	if len(out) == 0 {
		out = append(out, ValidationError{Code: "constraint", Path: "", Message: err.Error()})
	}
	return out
}

func collectVE(e *jsonschema.ValidationError, out *[]ValidationError) {
	if len(e.Causes) == 0 {
		*out = append(*out, ValidationError{
			Code:    classify(e.Error()),
			Path:    instancePath(e),
			Message: e.Error(),
		})
		return
	}
	for _, c := range e.Causes {
		collectVE(c, out)
	}
}

func instancePath(e *jsonschema.ValidationError) string {
	// jsonschema v6 exposes InstanceLocation as []string
	if loc := e.InstanceLocation; len(loc) > 0 {
		return "/" + strings.Join(loc, "/")
	}
	return ""
}

func classify(msg string) string {
	lower := strings.ToLower(msg)
	switch {
	case strings.Contains(lower, "required"):
		return "required"
	case strings.Contains(lower, "expected") && strings.Contains(lower, "type"):
		return "type"
	case strings.Contains(lower, "type"):
		return "type"
	case strings.Contains(lower, "format") || strings.Contains(lower, "pattern"):
		return "format"
	default:
		return "constraint"
	}
}

func validateNativeEmbedded(doc any) []ValidationError {
	root, ok := doc.(map[string]any)
	if !ok {
		return nil
	}
	imps, _ := root["imp"].([]any)
	var errs []ValidationError
	for i, raw := range imps {
		imp, _ := raw.(map[string]any)
		if imp == nil {
			continue
		}
		native, _ := imp["native"].(map[string]any)
		if native == nil {
			continue
		}
		req, _ := native["request"].(string)
		if req == "" {
			continue
		}
		var inner any
		if err := json.Unmarshal([]byte(req), &inner); err != nil {
			errs = append(errs, ValidationError{
				Code:    "native",
				Path:    fmt.Sprintf("/imp/%d/native/request", i),
				Message: "native.request is not JSON: " + err.Error(),
			})
			continue
		}
		if err := nativeSch.Validate(inner); err != nil {
			for _, ve := range mapSchemaError(err) {
				ve.Code = "native"
				ve.Path = fmt.Sprintf("/imp/%d/native/request%s", i, ve.Path)
				ve.Message = "native.request: " + ve.Message
				errs = append(errs, ve)
			}
		}
	}
	return errs
}

File: sdk/rust/src/validate.rs
//! Unified ValidationResult and BidRequest / BidResponse validators.

use jsonschema::Validator as JsValidator;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::OnceLock;

const OPENRTB: &str = include_str!("../schemas/openrtb.schema.json");
const BID_REQUEST: &str = include_str!("../schemas/bid-request.schema.json");
const BID_RESPONSE: &str = include_str!("../schemas/bid-response.schema.json");
const NATIVE: &str = include_str!("../schemas/native.schema.json");

/// Unified validation outcome (HTTP 400 body when `ok` is false).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ValidationResult {
    pub ok: bool,
    pub errors: Vec<ValidationError>,
}

/// One failing validation check.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ValidationError {
    pub code: String,
    pub path: String,
    pub message: String,
}

impl ValidationResult {
    pub fn ok_result() -> Self {
        Self {
            ok: true,
            errors: vec![],
        }
    }

    pub fn fail(errors: Vec<ValidationError>) -> Self {
        Self { ok: false, errors }
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("ValidationResult serializes")
    }

    pub fn to_json_bytes(&self) -> Vec<u8> {
        serde_json::to_vec(self).expect("ValidationResult serializes")
    }
}

fn request_validator() -> &'static JsValidator {
    static V: OnceLock<JsValidator> = OnceLock::new();
    V.get_or_init(|| compile(BID_REQUEST, "bid-request"))
}

fn response_validator() -> &'static JsValidator {
    static V: OnceLock<JsValidator> = OnceLock::new();
    V.get_or_init(|| compile(BID_RESPONSE, "bid-response"))
}

fn native_validator() -> &'static JsValidator {
    static V: OnceLock<JsValidator> = OnceLock::new();
    V.get_or_init(|| compile(NATIVE, "native"))
}

fn compile(schema_str: &str, label: &str) -> JsValidator {
    let schema: Value = serde_json::from_str(schema_str).unwrap_or_else(|e| {
        panic!("invalid {label} schema: {e}");
    });
    // Register sibling schemas for $ref resolution via retrieving
    let openrtb: Value = serde_json::from_str(OPENRTB).expect("openrtb schema");
    let resource = jsonschema::Resource::from_contents(openrtb).expect("openrtb resource");

    jsonschema::options()
        .with_resource(
            "https://github.com/oakrtb/openrtb/schema/jsonschema/openrtb.schema.json",
            resource.clone(),
        )
        .with_resource("openrtb.schema.json", resource)
        .build(&schema)
        .unwrap_or_else(|e| panic!("compile {label}: {e}"))
}

/// Validate BidRequest JSON bytes.
pub fn validate_bid_request(data: &[u8]) -> ValidationResult {
    validate(data, request_validator(), true)
}

/// Validate BidResponse JSON bytes.
pub fn validate_bid_response(data: &[u8]) -> ValidationResult {
    validate(data, response_validator(), false)
}

fn validate(data: &[u8], validator: &JsValidator, check_native: bool) -> ValidationResult {
    let doc: Value = match serde_json::from_slice(data) {
        Ok(v) => v,
        Err(e) => {
            return ValidationResult::fail(vec![ValidationError {
                code: "parse".into(),
                path: String::new(),
                message: e.to_string(),
            }]);
        }
    };

    let mut errors: Vec<ValidationError> = validator
        .iter_errors(&doc)
        .map(|e| ValidationError {
            code: classify(&e.to_string()),
            path: pointer_path(e.instance_path.to_string()),
            message: e.to_string(),
        })
        .collect();

    if check_native {
        errors.extend(validate_native_embedded(&doc));
    }

    if errors.is_empty() {
        ValidationResult::ok_result()
    } else {
        ValidationResult::fail(errors)
    }
}

fn pointer_path(path: String) -> String {
    if path.is_empty() || path == "/" {
        return path;
    }
    if path.starts_with('/') {
        path
    } else {
        format!("/{path}")
    }
}

fn classify(msg: &str) -> String {
    let lower = msg.to_ascii_lowercase();
    if lower.contains("required") {
        "required".into()
    } else if lower.contains("type") {
        "type".into()
    } else if lower.contains("format") || lower.contains("pattern") {
        "format".into()
    } else {
        "constraint".into()
    }
}

fn validate_native_embedded(doc: &Value) -> Vec<ValidationError> {
    let mut errors = Vec::new();
    let Some(imps) = doc.get("imp").and_then(|v| v.as_array()) else {
        return errors;
    };
    for (i, imp) in imps.iter().enumerate() {
        let Some(req) = imp
            .get("native")
            .and_then(|n| n.get("request"))
            .and_then(|r| r.as_str())
        else {
            continue;
        };
        let inner: Value = match serde_json::from_str(req) {
            Ok(v) => v,
            Err(e) => {
                errors.push(ValidationError {
                    code: "native".into(),
                    path: format!("/imp/{i}/native/request"),
                    message: format!("native.request is not JSON: {e}"),
                });
                continue;
            }
        };
        for e in native_validator().iter_errors(&inner) {
            errors.push(ValidationError {
                code: "native".into(),
                path: format!(
                    "/imp/{i}/native/request{}",
                    pointer_path(e.instance_path.to_string())
                ),
                message: format!("native.request: {e}"),
            });
        }
    }
    errors
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;
    use walkdir::WalkDir;

    fn repo_root() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../..")
            .canonicalize()
            .unwrap()
    }

    #[test]
    fn examples_pass() {
        let root = repo_root().join("examples");
        for entry in WalkDir::new(&root).into_iter().filter_map(|e| e.ok()) {
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) != Some("json") {
                continue;
            }
            let data = fs::read(path).unwrap();
            let parent = path.parent().unwrap().file_name().unwrap().to_string_lossy();
            let result = if parent == "bid-request" {
                validate_bid_request(&data)
            } else {
                validate_bid_response(&data)
            };
            assert!(
                result.ok,
                "{} => {:?}",
                path.display(),
                result.errors
            );
        }
    }

    #[test]
    fn invalid_rejected() {
        let root = repo_root().join("testdata/invalid");
        for entry in WalkDir::new(&root).into_iter().filter_map(|e| e.ok()) {
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) != Some("json") {
                continue;
            }
            let data = fs::read(path).unwrap();
            let parent = path.parent().unwrap().file_name().unwrap().to_string_lossy();
            let result = if parent == "bid-request" {
                validate_bid_request(&data)
            } else {
                validate_bid_response(&data)
            };
            assert!(!result.ok, "{} should fail", path.display());
            assert!(!result.errors.is_empty());
        }
    }

    #[test]
    fn parse_error() {
        let result = validate_bid_request(b"{");
        assert!(!result.ok);
        assert_eq!(result.errors[0].code, "parse");
    }
}
